feat: document live migration from TimescaleDB to Timescale (#2897)
timescale/docs: migrate/live-migration/live-migration-from-postgres.md

This document provides detailed instructions to migrate data from your
PostgreSQL database to a Timescale instance with minimal downtime (on the order
of a few minutes) of your production applications, using the [live migration]
strategy. To simplify the migration, we provide you with a docker image
containing all the tools and scripts that you need to perform the live
migration.

You should provision a dedicated instance to run the migration steps from.
Ideally an AWS EC2 instance that's in the same region as the Timescale target
service. For an ingestion load of 10,000 transactions/s, and assuming that the
historical data copy takes 2 days, we recommend 4 CPUs with 4 to 8 GiB of RAM
and 1.2 TiB of storage.

<SourceTargetNote />

In detail, the migration process consists of the following steps:

1. Set up a target database instance in Timescale.
1. Prepare the source database for live migration.
1. Run the live migration docker image.
1. Validate the data in target database and use it as new primary.

<GettingHelp />

<StepOne />

## 2. Prepare the source database for the live migration

<LiveMigrationStep2 />

## 3. Run the live migration docker image

> First, set the database URIs as environment variables:

Reviewer: This instruction is too late. The previous steps refer to `$SOURCE`. Also, I would prefer it if we could provide an actual url as an example, instead of an empty string. Compare with the instructions in the note for dual-write and backfill: https://docs.timescale.com/migrate/latest/dual-write-and-backfill/dual-write-from-timescaledb/#3-set-up-schema-and-migrate-relational-data-to-target-database

<img width="632" alt="image" src="https://github.com/timescale/docs/assets/1117749/236fb9e5-62ee-4574-9041-4b82bf283201">

import DumpDatabaseRoles from "versionContent/_partials/_migrate_dual_write_dump_database_roles.mdx";
import DumpPreDataSourceSchema from "versionContent/_partials/_migrate_pre_data_dump_source_schema.mdx";
import DumpPostDataSourceSchema from "versionContent/_partials/_migrate_post_data_dump_source_schema.mdx";
import LiveMigrationStep2 from "versionContent/_partials/_migrate_live_migration_step2.mdx";

# Live migration from PostgreSQL database with pgcopydb

> This document provides detailed instructions on how to perform migration
> from your production PostgreSQL database to a Timescale instance in a way
> that requires minimal downtime (in the order of few minutes) on your
> production applications.
> 
> The migration process will be carried out using the [live migration] strategy. You
> will use a live migration image that streamlines the process of replicating live
> transactions from your source database to Timescale.

Reviewer: ```suggestion
This document provides detailed instructions to migrate data from your
PostgreSQL database to a Timescale instance with minimal downtime (on the order
of a few minutes) of your production applications, using the [live migration]
strategy. To simplify the migration, we provide you with a docker image
containing all the tools and scripts that you need to perform the live
migration.
```

You should provision a dedicated instance to run the migration steps from.
Ideally an AWS EC2 instance that's in the same region as the Timescale target
service. For an ingestion load of 10,000 transactions/s, and assuming that the
historical data copy takes 2 days, we recommend 4 CPUs with 4 to 8 GiB of RAM
and 1.2 TiB of storage.

<SourceTargetNote />

In detail, the migration process consists of the following steps:

1. Set up a target database instance in Timescale.
1. Prepare the source database for live migration.
1. Run the live migration image.
1. Validate the data in target database and promote as new primary.

<GettingHelp />

<StepOne />

## 2. Prepare the source database for the live migration

<LiveMigrationStep2 />

> ## 3. Run the live migration image

Reviewer: ```suggestion
## 3. Run the live migration docker image
```

First, set the database URIs as environment variables:

```sh
export SOURCE=""
export TARGET=""
```

> Next, download and run the live migration image:

Reviewer: ```suggestion
Next, download and run the live migration docker image:
```

```sh
docker run --rm -it \
  -e PGCOPYDB_SOURCE_PGURI=$SOURCE \
  -e PGCOPYDB_TARGET_PGURI=$TARGET \

>   timescale/live-migration:0.0.1

Reviewer: The suggested tag doesn't exist.
```suggestion
  timescale/live-migration:main
```

```

This command may take a while to complete. It’s recommended to run it in the
background or use terminal multiplexers like `screen` or `tmux`.

The command will take a snapshot of your source database and migrate the schema

> the target database. Post migrating the schema, it will prompt to create

Reviewer: ```suggestion
the target database. After migrating the schema, it will prompt you to create
```

hypertables in the target database.

> ```sh

Reviewer: Stray code block
```suggestion
```

Now, let's transform your regular tables into hypertables.

Execute the following command for each table you want to convert on the target DB:
    `SELECT create_hypertable('<table name>', '<time column name>', chunk_time_interval => INTERVAL '<chunk interval>')`

> ```

Reviewer: stray
```suggestion

```

Ideally, tables that contain time-series data should be converted to Hypertables.
You need to run `create_hypertable()` for each table that you want to convert to
a Hypertable in the target database. For more information, see [Hypertable docs].

Once you have finished creating Hypertables, you need to signal "continue" to proceed.

> You can do it either by pressing `c` key or giving a `SIGUSR1` signal,

Reviewer: I find it confusing that we provide both options "pressing `c` or sending `SIGUSR1`. Why do we have both options? Can we just skip mentioning `SIGUSR1`?

Reviewer 2: +1, since we advocate to use docker in interactive mode(-it), there is no need to mention about signals.

as mentioned in your log.

Next, the live migration image will migrate the existing data in the source database
to target database and start streaming live transactions (live replay) received on
the source side to the target. During this process, it will display the lag between
the source and target databases in terms of WAL offset size.

```sh
[WATCH] Source DB - Target DB => 126MB
```

When the lag between the source and target database is less than 30 megabytes,
it will start `ANALYZE` on the target database. This updates statistics in the
target database, which is necessary for optimal querying performance in the
target database. Wait for `ANALYZE` to complete.

<Highlight type="note">
Application downtime begins here.
</Highlight>

> When you’re _ready to take downtime_ (i.e., stop normal operations) on your
> production applications that are dependent on your source database and the
> _lag between the databases is below 30 megabytes_, you should stop your
> applications that perform DML queries on your source database. This is
> the downtime phase and will last until you have completed the validation
> step (4). Be sure to go through the validation step (4) and prepare the
> validation queries before you enter the downtime phase to keep the overall
> downtime minimal.

Reviewer: ```suggestion
Once the lag between the databases is below 30 megabytes, and you're ready to
take your applications offline, stop all applications which are writing to the
source database. This is the downtime phase and will last until you have
completed the validation step (4). Be sure to go through the validation step
(4) and prepare the validation queries before you enter the downtime phase to
keep the overall downtime minimal.
```

> Stopping DML queries on your source database allows the live migration
> image to make the target database even with the source. This will be
> evident when the lag outputs 0 megabytes.

Reviewer: ```suggestion
Stopping writes to the source database allows the live migration process to
finish replicating data to the target database. This will be evident when the
replication lag reduces to 0 megabytes.
```

> When the lag between the two databases shows 0 megabytes, wait for a
> few minutes and then provide the signal to stop the live replay by
> pressing key `s` or giving a `SIGUSR1` signal, as per your log.
> You should see `Stopped` message.

Reviewer: ```suggestion
Once the replication lag is 0, wait for a few minutes and then provide the
signal to stop the live replay by pressing key `s` or giving a `SIGUSR1`
signal, as per your log.
You should see a `Stopped` message.
```

```sh
[WATCH] Source DB - Target DB => 0MB. Press 's' (and ENTER) to stop live-replay
Stopped
```

Now, compare tables in the target against those in the source to ensure
that both have the same data.

Once you are confident that the tables in the source are identical to the
target, you need to signal “continue”. This is done either by pressing
key `c` or giving a `SIGUSR1` signal, as mentioned in your log.

```sh
[ACTION NEEDED] Now, you should check the integrity of your data. Once you are confident, press 'c' (and ENTER) to continue
Continuing ...
Syncing last LSN in Source DB to Target DB ...
```

The live migration image will continue the remaining work that includes
migrating sequences and running policies. You should see the following
message if all the mentioned steps were successful.

```sh
Migration successfully completed
```

> ## 4. Validate the data in target database and promote as new primary

Reviewer: I don't really like the word "promote" here, it implies doing something to the database, which we're not doing.
```suggestion
## 4. Validate the data in target database and use it as new primary
```

Now that all data has been migrated, the contents of both databases should
be the same. How exactly this should best be validated is dependent on
your application. You could compare the number of rows or an aggregate of
columns to validate that the target database matches with the source.

<Highlight type="note">
Application downtime ends here.
</Highlight>

> Once you are confident with the validation perspective, the final step is
> to promote the target database as your new production database by using
> the target database's URI in your production application.

Reviewer: ```suggestion
Once you are confident with the data validation, the final step is to configure
your applications to use the target database.
```
